Source/DafnyCore/AST/Members/Function.cs

```diff
@@ -359,6 +360,17 @@ public enum CoCallClusterInvolvement {
   protected override bool Bodyless => Body == null;
   protected override string TypeName => "function";
 
+  public static readonly Option<bool> AllowExternalFunction = new("--allow-external-function",
+    "When false, a warning is emitted for functions marked with the {:extern} attribute. Dafny warns about such functions because a function can always be implemented in pure Dafny code. Using an externally implemented function may improve performance or reduce development time, but means less code is verified");
```

Comment:
I'd change the second sentence to "Dafny warns about such functions because external code rarely implements pure functions, and usually has effects of some sort."

One thing I'm starting to think is that we might want per-function attributes to disable this warning, rather than a global option. If you have a code base that already has a few external functions, which you've carefully vetted to actually be functions, and then you add a new one without careful consideration, it would be useful to have a warning about that.

Reply:
"external code rarely implements pure functions, and usually has effects of some sort." I'm not a fan of that. What do you mean by rarely? You can find entirely libraries (or even languages?) with pure code.

I think the argument should be that even though the external code could be pure, it would be better if you use Dafny code because then it will be verified.

Reply:
> When someone decides to use external code for something, it takes a good deal of extra work

What do you mean? A user might use external code to do less work, since then they don't have to implement something in Dafny.

> If we have this warning at all, I think it should be because there's a soundness risk.

You're introducing a soundness risk by using external code where pure Dafny would have sufficed. Why would you not warn about that?

Reply:
> I'm not comfortable with defining our diagnostics based on our gut feeling estimates of risks. I'm suggesting a warning here because functions have a property that methods do not.

Any external code is a risk, and can introduce soundness issues, and I'm all in favor of warning about that. But I don't like the idea of telling our users "You should implement everything in Dafny that you possibly can" because that's their decision to make, based on whatever tradeoffs they're dealing with that we have no way of knowing about. It's just not a friendly message.

Reply:
> "You should implement everything in Dafny that you possibly can" because that's their decision to make, based on whatever tradeoffs they're dealing with that we have no way of knowing about. It's just not a friendly message.

I agree. I tried to keep the warning description informative and unopinionated, but maybe I did not do a good job at that. I'll try to update the messaging a little. LMK if you have suggestions.

I think if the trade-off leans strongly enough towards Dafny, in the sense that most users won't have to use `--allow-external-function`, then it makes sense to have this warning. I'm also hopeful that Dafny will get to that point in the future, and I would be OK with already having this warning even if we're not there yet. It's a way of pushing users to use more Dafny in their software, which gives us more feedback if there are performance issues or important libraries missing from the ecosystem.

You could also argue that this should be a suggestion not a warning, even in the late game, but I feel when a user chooses to introduce a correctness risk in their application, it's good if Dafny requires them to make that choice explicit.

Reply:
I like how it's turned out. I think having a warning on any external contracts makes sense.

Reply:
Thanks